### ISSUE TITLE
cdc: key_column option unsupported with CDC query

### ISSUE THREAD

**Issue by miretskiy** [cockroachdb/cockroach#114196]
https://github.com/cockroachdb/cockroach/pull/98806 added support for specifying custom
`key_column` -- however, this neglected to test CDC query functionality.
When using key_column with CDC query, the changefeed will fail with:
`retryable error: No column with name <name> in this row`

The problem is that CDC queries construct event rows using [projection](https://github.com/cockroachdb/cockroach/blob/c3d4c677b54d7d737cc0071b5bc1ffab96ae8004/pkg/ccl/changefeedccl/cdcevent/projection.go#L31-L31) and when copying EventDescriptor, the newly added [name->column id map](https://github.com/cockroachdb/cockroach/blob/c3d4c677b54d7d737cc0071b5bc1ffab96ae8004/pkg/ccl/changefeedccl/cdcevent/event.go#L254-L254) from the event descriptor is not copied.

Informs https://github.com/cockroachlabs/support/issues/2697

Jira issue: CRDB-33389

**Comment by wenyihu6** [comment]
After some testing, these are things that work/don't work with custom key columns so far: 
TLDR: it has to be an actual column that exists on the table - not something created in CDC queries or created afterwards 
1. using a key column that user created in CDC queries
```
CREATE CHANGEFEED WITH key_column='double_b', unordered AS SELECT concat(b, c) AS double_b FROM foo
```
2. selecting on a virtual column does not work
```
CREATE TABLE foo (a INT PRIMARY KEY, b STRING, c STRING, d STRING AS (concat(b, c)) VIRTUAL)
CREATE CHANGEFEED WITH key_column='d', unordered AS SELECT d FROM foo 
```
4. selecting on a hidden column works
```
CREATE TABLE foo (a INT PRIMARY KEY, b STRING NOT VISIBLE)
CREATE CHANGEFEED WITH key_column='b', unordered AS SELECT b FROM foo
```
5. selecting on cdc_prev or cdc_is_delete doesn't seem to work, and cdc_is_delete is not recognized as a function either when used in CDC queries 
```
CREATE CHANGEFEED WITH key_column='b', unordered AS SELECT cdc_is_delete() FROM foo
```
6. selecting on `crdb_internal_mvcc_timestamp` works
```
CREATE CHANGEFEED WITH key_column='crdb_internal_mvcc_timestamp', unordered AS SELECT crdb_internal_mvcc_timestamp FROM foo WHERE crdb_internal_mvcc_timestamp > 0
```
8. using CDC queries with event_op works 
```
CREATE CHANGEFEED WITH key_column='b', unordered AS SELECT * FROM foo WHERE event_op() = 'delete'
```

@miretskiy does anything above come as a surprise and we should fix?

**Comment by miretskiy** [comment]
> selecting on cdc_prev or cdc_is_delete doesn't seem to work, and cdc_is_delete is not recognized as a function either when used in CDC queries
CREATE CHANGEFEED WITH key_column='b', unordered AS SELECT cdc_is_delete() FROM foo

This is "old" style function -- use event_op() = deleted to determine if it's deletion.
